Address #5912: google-two-tier example incomplete

mcalthrop opened the pull request:
This addresses #5912.

After these modifications to the `google-two-tier` example have been made, it now works when instructions are followed.

I made the following changes:
* updated the `README` with additional instructions on how to run this example, including private/public key generation
* changed `machine_type` from `n1-standard-1` to `f1-micro`
* updated the ubuntu image to one that makes use of `ssh-keys` instead of `sshKeys`
* use a `provisioner` instead of `startup-script` &ndash; the latter neither shows errors, nor causes `terraform` to fail on error; the former does
* added `scripts/install.sh`, which is used by the `provisioner`
* run SSH install script as `root` instead of `ubuntu` (the script runs `apt-get`, which requires root privileges)

If all defaults are used, the minimum command for running this example is:

```sh
$ terraform apply -var="project_name=your-google-compute-project-name"
```

sparkprime commented:
It's sad that you have to use a provisioner to get errors, because that means a different config is required to when autoscaling is used.  I'd love to see an error stream from startup scripts to address this.  Currently I think you can get it from the instance's serial port output, but that contains a lot of other stuff folded in and it's hard for a tool like Terraform to separate.

sparkprime commented:
Is that ssh key stuff needed specifically for provisioners?  You should be able to ssh to the instance by adding your ssh key to the project wide metadata, which can be done in the cloud console.  That's easier than generating keys IMO.  Otherwise, ssh keys can also be distributed on-demand by running gcloud ssh, but this won't help with the provisioner.

mcalthrop commented:
@sparkprime: thanks for your replies (received just after I went to bed! I'm in the UK).

Re. errors: there may be another way to trap errors and fail on error, but this was the only way I found to do it. Bear in mind this is my first foray into Terraform, so there may be a better way!

Re. SSH keys: yes, they are required for the provisioners (I used the existing code as a starting point). I'll look into adding SSH keys at the project level, and update the PR if I find a way.

mcalthrop commented:
Update: I've tried several different things, and so far nothing works.

I wonder if it's because you have to specify a user for each project-wide SSH key (I even tried adding the same public key twice, once for user `root` and once for user `ubuntu`).

The errors I got were:
```
Error applying plan:

3 error(s) occurred:

* Failed to upload script: Error reading script: EOF
* Failed to upload script: Error reading script: EOF
* Failed to upload script: Error reading script: EOF
```
and also:
```
Error applying plan:

3 error(s) occurred:

* ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
* ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
* ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
```

Any suggestions?

sparkprime commented:
Yeah I don't think there's a better way to get errors out right now than using provisioner, not just in Terraform but generally for GCE.

The intended way to use keys is that you associate your public key (either your regular key or some new one, in which case you need to use ssh -i) with the username you intend to log in with, which is typically the same username you use on all your personal machines.  If many people are added to a project, they all add their own ssh keys and these are managed centrally, with agents to dynamically update the instance configuration.  This is radically different to how AWS does it, which is that there is one username for everybody (e2c-user or ubuntu I think).  So the question is what username does the provisioner ssh to?  And should it be configured to use $USER for that?  I'm not sure if sshing as root is allowed.  You can always test all of this with ssh from outside Terraform.

mcalthrop commented:
Thanks for the keys explanation... pretty much how I understood it.

The code in the PR definitely works, and that sets the user as `root` for both the `file` provisioner and the `remote-exec` provisioner. So I tried uploading my public key at the _project_ level, for the `root` user. I simplified the `main.tf` file to remove the `metadata` section, and made the provisioners look like this:

```
  provisioner "file" {
    source = "${var.install_script_src_path}"
    destination = "${var.install_script_dest_path}"
    connection {
      agent = false
    }
  }

  provisioner "remote-exec" {
    connection {
      agent = false
    }
    inline = [
      "chmod +x ${var.install_script_dest_path}",
      "${var.install_script_dest_path} ${count.index}"
    ]
  }
```

I then ran `terraform destroy`, followed by `terraform apply`, and got errors:

```
google_compute_instance.www.0: Provisioning with 'file'...
google_compute_instance.www.1: Provisioning with 'file'...
google_compute_instance.www.2: Provisioning with 'file'...
Error applying plan:

3 error(s) occurred:

* ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
* ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
* ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
```

I added `type = "ssh"` to both of the `connection` blocks, and got the same errors.

I also added the same public key at the project level for the `root` user – and still got errors (my internet connection is very flaky at the moment, so am having trouble running it).

sparkprime commented:
Maybe what is happening is the provisioner needs to be told to use $HOME/.ssh/id_dsa.pub or whatever it is, otherwise they don't provide a key at all.

sparkprime commented:
Alright what you're doing seems reasonable then. I'll merge.
